File: Problem/By section/Binary Search/Problem 153/problem_153.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    // Time: O(log(n)), Space: O(1)
    int findMin(vector<int>& nums) {
        int size = nums.size();
        int left = 0, right = size - 1;
        int mid;

        while(left < right) {
            mid = left + (right - left) / 2;
            
            if(nums[mid] > nums[right]) {
                left = mid + 1;
            }else if(nums[mid] < nums[right]) {
                right = mid;
            }
        }

        return nums[left];
    }
};

int main(int argc, char** argv) {
    Solution s;
    vector<int> input = {3, 1, 2};
    cout << s.findMin(input) << endl;
    return 0;
}

File: Problem/By section/Binary Search/Problem 153/problem_153.java
class Solution {
    // Time: O(logn), Space: O(1)
    public int findMin(int[] nums) {
        if(nums.length <= 1) return nums[0];
        
        return bin_search(nums, 0, nums.length - 1);
    }
    
    // Successfully prune half of the branch
    // because we dont care about the increasing part
    public int bin_search(int[] nums, int l, int r){
        if(l == r) return nums[l];
        
        int middle = l + (r - l) / 2;
        
        if(nums[middle] < nums[r]){
            return bin_search(nums, l, middle);
        }else{
            return bin_search(nums, middle + 1, r);
        }
    }


    //     public int bin_search(int[] nums, int l, int r){
    //     if(l <= r){
             
    //         int middle = l + (r - l) / 2;
            
    //         if(nums[middle - 1] > nums[middle] && nums[middle + 1] > nums[middle]){
    //             return nums[middle];
    //         }else{
    //             return Math.min(bin_search(nums, l, middle - 1), bin_search(nums, middle + 1, r));
    //         }
    //     }else{
    //         return Integer.MAX_VALUE;
    //     }
    // }
}
